homework/homework11/library.py

```diff
@@ -0,0 +1,92 @@
+"""Library.
+This program shows whether the user can borrow a reserved book.
+"""
+
+
+class Books:
+    """Initialize book attributes"""
+    def __init__(self, book_name, author, isbn, page_count):
+        self.book_name = book_name
+        self.author = author
+        self.isbn = isbn
+        self.page_count = page_count
+        self.reserved = False
+        self.busy = False
+
+    def get_description_book(self):
+        """Get description of the book"""
+        descr = (f'book_name: {self.book_name}, '
+                 f'author: {self.author}, '
+                 f'isbn: {str(self.isbn)}, '
+                 f'page_count: {str(self.page_count)}')
+        print(f'Book Description: {descr}')
+
+    def reservation(self):
+        """This method allows the user to reserve a book if it is not busy"""
+        if not self.busy:
+            self.reserved = True
+            return "Book has been successfully reserved"
+        return "The book is busy. You cannot book the book"
+
+    def receiving(self):
+        """This method allows a user to borrow a book if it is not reserved
+        or busy by another user."""
+        if not self.busy and not self.reserved:
+            self.busy = True
+            return "The book is free. You can take the book"
+        if self.reserved:
+            return "The book is reserved. You can't take the book"
+        return "The book is already taken"
+
+    def returning(self):
+        """This method allows the user to return the book"""
+        if self.busy:
+            self.reserved = False
+            self.busy = False
+            return "The book has been returned"
+        if self.reserved:
+            self.reserved = False
+            return "The book reserve has been canceled"
+        return "The book was free"
+
+
+class User:
+    """Initialize user actions"""
+    @staticmethod
+    def take_book(book):
+        """This method allows the user to take the book"""
+        return book.receiving()
+
+    @staticmethod
+    def reserve_book(book):
+        """This method allows the user to reserve a book"""
+        return book.reservation()
+
+    @staticmethod
+    def return_book(book):
+        """This method allows the user to return the book"""
+        return book.returning()
+
+
+book_1 = Books('The Lord of the Rings', 'J.R.R. Tolkien',
+               9785171358136, 1120)
+book_1.get_description_book()
+
+user_1 = User()
+user_2 = User()
+
+
+print(user_1.take_book(book_1))
+print(user_2.take_book(book_1))
+print()
+
+print(user_1.return_book(book_1))
```

Comment:
user_2 can not return book, that was taken by another user
```
print(user_1.take_book(book_1))
print(user_2.return_book(book_1))
```

Reply:
@Stml89, Is the problem in the second user's action or in the code? Can I change the second user's action here, e.g. also take the book?

Reply:
print(user_1.return_book(book_1))
print(user_2.reserve_book(book_1))

print(user_1.take_book(book_1))
print(user_2.return_book(book_1))

here user 1 can't take the book because it was previously reserved and after user 2 canceled the book reservation user 1 can take it)
it seems to be logically ok, or did I not understand you correctly?)

Reply:
My point is that if user1 reserved a book, how can user2 return it?

Reply:
@Stml89 ok, understand. I replaced the action for user2. Now, users can be take and return book correctly)

Reply:
Nope, the idea was to check that if user1 took a book, than another user2 can not return it:
```
book = Book(book_name="The Hobbit", author="Books by J.R.R. Tolkien", num_pages=400, isbn="0006754023")

vasya = Reader("Vasya")
petya = Reader("Petya")


vasya.reserve_book(book)
petya.reserve_book(book) <-- User can not reserve a book
vasya.cancel_reservation(book)

petya.reserve_book(book)
vasya.get_book(book) <-- User can not get a book
petya.get_book(book)
vasya.return_book(book) <-- User can not return a book
petya.return_book(book)

vasya.get_book(book)
```

Reply:
Phew, guess I fixed it, check it out please)